routing/index_graph.cpp

```diff
@@ -0,0 +1,197 @@
+}
+
+void IndexGraph::Export(vector<Joint> const & joints)
+{
+  m_fsegIndex.Export(joints);
+  BuildJoints(joints.size());
+}
+
+JointId IndexGraph::InsertJoint(FSegId fseg)
+{
+  JointId const existId = m_fsegIndex.GetJointId(fseg);
+  if (existId != kInvalidJointId)
+    return existId;
+
+  JointId const jointId = m_jointOffsets.size();
+  m_jointOffsets.emplace_back(JointOffset(m_fsegs.size(), m_fsegs.size() + 1));
+  m_fsegs.emplace_back(fseg);
+  m_fsegIndex.AddJoint(fseg, jointId);
+  return jointId;
+}
+
+vector<FSegId> IndexGraph::RedressRoute(vector<JointId> const & route) const
+{
+  vector<FSegId> fsegs;
+  if (route.size() < 2)
+  {
+    if (route.size() == 1)
+      fsegs.emplace_back(GetFSeg(route[0]));
+    return fsegs;
+  }
+
+  fsegs.reserve(route.size() * 2);
+
+  for (size_t i = 0; i < route.size() - 1; ++i)
+  {
+    JointId const prevJoint = route[i];
+    JointId const nextJoint = route[i + 1];
+
+    auto const & pair = FindSharedFeature(prevJoint, nextJoint);
+    if (!pair.first.IsValid())
+    {
+      fsegs.clear();
+      return fsegs;
+    }
+
+    if (i == 0)
+      fsegs.push_back(pair.first);
+
+    uint32_t const featureId = pair.first.GetFeatureId();
+    uint32_t const segFrom = pair.first.GetSegId();
+    uint32_t const segTo = pair.second.GetSegId();
+
+    ASSERT_NOT_EQUAL(segFrom, segTo, ());
+
+    if (segFrom < segTo)
+    {
+      for (uint32_t seg = segFrom + 1; seg < segTo; ++seg)
+        fsegs.push_back({featureId, seg});
+    }
+    else
+    {
+      for (uint32_t seg = segFrom - 1; seg > segTo; --seg)
+        fsegs.push_back({featureId, seg});
+    }
+
+    fsegs.push_back(pair.second);
+  }
+
+  return fsegs;
+}
+
+FSegId IndexGraph::GetFSeg(JointId jointId) const
+{
+  return m_fsegs[GetJointOffset(jointId).Begin()];
+}
+
+inline JointOffset const & IndexGraph::GetJointOffset(JointId jointId) const
+{
+  ASSERT_LESS(jointId, m_jointOffsets.size(), ("JointId out of bounds"));
+  return m_jointOffsets[jointId];
+}
+
+pair<FSegId, FSegId> IndexGraph::FindSharedFeature(JointId jointId0, JointId jointId1) const
+{
+  JointOffset const & offset0 = GetJointOffset(jointId0);
+  JointOffset const & offset1 = GetJointOffset(jointId1);
+
+  for (size_t i = offset0.Begin(); i < offset0.End(); ++i)
+  {
+    FSegId const & fseg0 = m_fsegs[i];
+    for (size_t j = offset1.Begin(); j < offset1.End(); ++j)
+    {
+      FSegId const & fseg1 = m_fsegs[j];
+      if (fseg0.GetFeatureId() == fseg1.GetFeatureId())
+        return make_pair(fseg0, fseg1);
+    }
+  }
+
+  LOG(LERROR, ("Can't find shared feature for joints", jointId0, jointId1));
+  return make_pair(FSegId::MakeInvalid(), FSegId::MakeInvalid());
+}
+
+void IndexGraph::BuildJoints(uint32_t jointsAmount)
+{
+  // +2 is reserved space for start and finish
+  m_jointOffsets.reserve(jointsAmount + 2);
+  m_jointOffsets.resize(jointsAmount, {0, 0});
+
+  m_fsegIndex.ForEachRoad([this](uint32_t featureId, RoadJointIds const & road) {
```

Comment:
m_fsegIndex.ForEachRoad([this](uint32_t /* featureId */, RoadJointIds const & road)